psi4/driver/driver.py

```diff
@@ -1932,35 +1932,137 @@ def gdma(wfn, datafile=""):
+    :type debug: boolean
+    :param debug: returns a dictionary to aid with debugging
 
     Notes
     -----
     * A description of the FCHK format is http://wild.life.nctu.edu.tw/~jsyu/compchem/g09/g09ur/f_formchk.htm
     * The allowed headers for methods are general and limited, i.e., "Total SCF|MP2|CI|CC Density",
-      so "CC" is always used for the post-HF case.
+      PSI4 will try to find the right one for the current calculation.
+    * Not all theory modules in PSI4 are compatible with the FCHK writer. Currently available and tested are:
+      All SCF/DFT variants (if gradient avaiable), DF-RHF-MP2 (dfmp2 module), CONV-RHF/UHF-CC methods (ccenergy module). No UHF-MP2 densities are available.
+      A warning will be printed if a theory module is not supported.
 
     :examples:
 
     >>> # [1] FCHK file for DFT calculation
     >>> E, wfn = energy('b3lyp', return_wfn=True)
     >>> fchk(wfn, 'mycalc.fchk')
 
+    >>> # [2] FCHK file for correlated densities
+    >>> E, wfn = gradient('ccsd', return_wfn=True)
+    >>> fchk(wfn, 'mycalc.fchk')
+
     """
+    #NOTE: This function should be overhauled once all modules set Ca/Cb, Da/Db in the same manner.
+
+    # QC_MODULE overview for FCHK
+    # OCC: (includes both occ/dfocc modules
+    #  wfn.D() = SCF
+    #  wfn.C() = SCF
+    #  refwfn.D() = SCF
+    # -> not allowed, graceful exit
+    # DCT:
+    #  wfn.D() = SCF
+    #  wnf.C() = DCT
+    #  refwfn.D() = SCF
+    # -> allowed, but FCHK entries are switched around
+    # CCENERGY
+    #  wfn.D() = CC
+    #  wfn.C() = SCF 
+    #  refwfn.D() = CC
+    # -> allowed, will recompute SCF density from wfn.C()
+    # DFMP2:
+    #  wfn.D() = MP2
+    #  wnf.C() = NO(MP2)
+    #  refwfn.D() = SCF 
+    # -> allowed, will take refwfn.D() for SCF density
+
+    exclude_list = ['OCC','DFOCC']
+    module_ = wfn.module().upper()
+    if (module_ in exclude_list):
+        core.print_out(f"THEORY MODULE {module_} is currently not supported by the FCHK writer.")
+        print(f"THEORY MODULE {module_} is currently not supported by the FCHK writer.")
+        return None
+
     fw = core.FCHKWriter(wfn)
+
+    # At this point we don't know the method name, so we try to search for it.
+    # idea: get the method from the variable matching closely the 'current energy'
+    varlist = core.scalar_variables()
+    current = varlist['CURRENT ENERGY']
+    del varlist['CURRENT ENERGY']
+
+    # delete problematic entries
+    delete_list = ['CURRENT ENERGY', 'CURRENT REFERENCE ENERGY']
+    for key in delete_list[:]:
+        if key in varlist:
+            del varlist[key]
+
+    # find closest matching energy
+    for (key, val) in varlist.items():
+        if (np.isclose(val, current,1e-10)):
+            method = key
+
+    # The labels this functions knows. Extend if needed.
+    labels = {
+    "HF" :" SCF Density",
+    "SCF":" SCF Density",
+    "DFT":" SCF Density",
```

Comment:
ROHF is missing, right?

Reply:
CUHF, too.

Reply:
Not sure what you mean. As a label? All `reference` options use the same `SCF TOTAL ENERGY` variable for the detection.
The "HF" line is currently not used but also does not hurt to have in case `UHF TOTAL ENERGY`or something like that is introduced.